opm/simulators/utils/PartiallySupportedFlowKeywords.cpp

```diff
@@ -27,55 +30,309 @@ template <>
 const PartiallySupportedKeywords<std::string>&
 partiallySupported()
 {
-    static const PartiallySupportedKeywords<std::string> partially_supported_keywords_strings = {
-        {
+   static const PartiallySupportedKeywords<std::string> partially_supported_keywords_strings = {
+         {
+            "BRINE",
+            {
+               {1,{false, allow_values<std::string> {}, "SALTS Multi-Component Brine model not supported"}}, // NO_NAME
```

Comment:
How should an empty list of allowed values be interpreted?

Reply:
An empty list implies that no values are accepted.

Reply:
Note some of the default values in the jason definitions are incorrect based on the commercial simulator values (TABDIMS for example). I have used the "correct" values so there will be some superfluous warning message until we amend the jason definitions. I can do this if you like, but l would like to have a quick discussion first.

Reply:
> An empty list implies that no values are accepted.

OK - I understand; another way of putting is: "This item must be defaulted"? Instead of outputting an empty list in this situation I would say a better solution would be to create another datastructure with all the keywords & item combinations which *must be defaulted*; that will require a minor update to the framework though.

When it comes to the `BRINE` keyword specifically that is a story of its own - there are some issues with the json definition there.

Reply:
> Note some of the default values in the jason definitions are incorrect based on the commercial simulator values (TABDIMS for example).

We can discuss - but I see no reason not to just go along and update.

Reply:
> OK - I understand; another way of putting it is: "This item must be defaulted"? Instead of outputting an empty list in this situation I would say a better solution would be to create another datastructure with all the keywords & item combinations which _must be defaulted_; that will require a minor update to the framework though.

From my end (a user of the framework) this sounds more complicated, I favour the current approach. Please keep in mind that the framework is not for listing supported options, it is for listing unsupported options. As long as it is easy to find the  meaning of an empty list (e.g., explaining the syntax at the top of the file), I am fine with it. If not, then my preference would rather be to use a symbol. For instance a star * instead of an empty list to indicate that all options are unsupported. I believe we should thrive for supporting the default option in any supported keyword, so that it can be taken for granted (not sure we are there now, but maybe?).

Reply:
> OK - I understand; another way of putting it is: "This item must be defaulted"? Instead of outputting an empty list in this situation I would say a better solution would be to create another datastructure with all the keywords & item combinations which must be defaulted; that will require a minor update to the framework though.

What about putting this in the jason file instead and try to consolidate everything in these files eventually. At the moment we have PartiallySupportedFlowKeywords.cpp and UnsupportedFlowKeywords.cpp. What do you think?

Also is it okay to add another parameter to the jason files, item=1, etc. to keep count of the number of variables, I would only do this for the keywords with lots of variables like TABDIMS and REGDIMS..

Reply:
> What about putting this in the jason file instead and try to consolidate everything in these files eventually. At the moment we have PartiallySupportedFlowKeywords.cpp and UnsupportedFlowKeywords.cpp. What do you think?

In many ways I would have liked to sprinkle information like this through the json scheme, the problem with that is that the json scheme does not know what the simulator supports, so that would be layering violation as some guy used to say.

> Also is it okay to add another parameter to the jason files, item=1, etc. to keep count of the number of variables

Yes - that is OK. They must be entered with correct json syntax - but will otherwise be ignored.

Reply:
> What about putting this in the jason file instead and try to consolidate everything in these files eventually. 

I am afraid this breaks with the intent of opm-simulators. The parser should not need to care what the simulators support, the parser is actually designed to support all keywords and options. I also fail to see how moving everything to the parser will make life easier for those who will maintain PartiallySupportedFlowKeywords.cpp and UnsupportedFlowKeywords.cpp.

Reply:
> If not, then my preference would rather be to use a symbol. For instance a star * instead of an empty list to indicate that all options are unsupported. I believe we should thrive for supporting the default option in any supported keyword, so that it can be taken for granted (not sure we are there now, but maybe?)

I like `*` or `1*`to indicate that all options are unsupported.

Reply:
> I like `*` or `1*`to indicate that all options are unsupported.

Great, sounds like a plan! But please do not hold this PR for it, we can easily update these occurrences from empty to stars once we get the improvement in the framework.

Reply:
Ehhh no - this is not a viable plan within the scope of the current framework. The current framework is based on C++ *code* - and that limits our ability to choose syntax ourselves. In order to allow syntax like `*`or `1*` to indicate "only default allowed" we will have to change to use data (i.e. strings of some kind) to configure this.

Reply:
> Ehhh no - this is not a viable plan within the scope of the current framework. The current framework is based on C++ _code_ - and that limits our ability to choose syntax ourselves. In order to allow syntax like `*`or `1*` to indicate "only default allowed" we will have to change to use data (i.e. strings of some kind) to configure this.

Ah, you are right. We need to think a bit more about this then. But great if we can go with the current solution in the meantime.

Reply:
What about a numeric constant -999.999 or -1.0E-20, for example

Reply:
> What about a numeric constant -999.999 or -1.0E-20, for example

Not sure about others, but to me an empty list is actually more intuitive then.

Reply:
> The current framework is based on C++ _code_ - and that limits our ability to choose syntax ourselves. In order to allow syntax like `*`or `1*` to indicate "only default allowed" we will have to change to use data (i.e. strings of some kind) to configure this.

Don't we already support function-like predicates in this position?  How about
```C++
struct default_only
{
    bool operator()(const std::string& x) const
    {
        return x == "*";
    }
};
```
or whatever we need to do to match defaulted values.  Then the specification would be something along the lines of
```C++
{1, {false, default_only{}, "SALTS ..."}}
```

Reply:
Yes we support function predicates, but the desire is to have a literal `*` in the code - not compare string data with `*`

Reply:
> Yes we support function predicates, but the desire is to have a literal `*` in the code

Right, but since that's not possible we need a semantically equivalent alternative that can be expressed in C++ source code.

> - not compare string data with `*`

What is the appropriate criterion then?  I'm starting to lean towards passing the `DeckItem` as an argument to the predicate rather than `DeckItem::get<T>()`.  That way, `default_only` (or however we choose to name it) can return
`item.defaultApplied(0)` instead of trying to match the value in an intelligent way.  The predicates that **do** need to look at the values can call `DeckItem::get<T>()` internally.

Reply:
The current semi-good solution is to pass `allow_values<T>{}` as an indirect way of saying anything but `*` in input is illegal (mind that this will fail to validate even input where the default value is entered explicitly).

The two alternatives I have thought of are:

1. Add a new datastructure with {keyword,item} pairs which *must be defaulted*.
2. Pass in `DeckItem&` instead of `DeckItem.get<T>(0)`

Alternative 2 will certainly give the most flexibility for this and possible future scenarios; however that will require immediate significant refactoring to the framework. I can discuss and assist on this - but I will not drive it in any direction.

Reply:
> The current semi-good solution is to pass `allow_values<T>{}` as an indirect[1] way of saying anything

It's rather a way of saying **nothing**.  The `allow_values<>` [predicate](https://github.com/OPM/opm-simulators/blob/c101ed10f4f69127311e71f32c8c6e9e28ef1367/opm/simulators/flow/KeywordValidation.hpp#L144-L147) matches `DeckItem::get<>(0)` against a list of permissible values.  If that list is empty, then any value at all will fail to match and the input will be deemed invalid.

I still think that passing the `DeckItem` as an argument is the better solution.  Note that any specification that uses `allow_values<>` will be automatically updated once we change the function call operator of that type.  We could even just add a new function call operator of the form
```C++
bool operator()(const DeckItem& item) const
{
    return (*this)(item.get<T>(0));
}
```
The downside is that we would need to include `DeckItem.hpp` into the `KeywordValidation.hpp` header.